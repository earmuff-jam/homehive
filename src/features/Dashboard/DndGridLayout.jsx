import { useState } from "react";
import { Responsive as ResponsiveGridLayout } from "react-grid-layout";

import { Stack, Typography } from "@mui/material";
import Widget from "src/features/Dashboard/Widget";

export default function DndGridLayout({
  widgets,
  setWidgets,
  handleRemoveWidget,
}) {
  const [activeWidget, setActiveWidget] = useState(null); // active widget for drag overlay

  const handleDragStart = (ev) => {
    const { active } = ev;
    const activeId = active.id.toString();
    const widget = widgets.find((w) => w.widgetID === activeId);
    setActiveWidget(widget);
  };

  const handleDragEnd = (ev) => {
    const { active, over } = ev; // active is current, over is replacing

    setActiveWidget(null);

    if (!over) return;
    if (active.id === over.id) return;

    const originalIdx = widgets.findIndex(
      (widget) => widget.widgetID === active.id.toString()
    );
    const newIdx = widgets.findIndex(
      (widget) => widget.widgetID === over.id.toString()
    );

    const updatedWidgets = arrayMove(widgets, originalIdx, newIdx);
    setWidgets(updatedWidgets);
  };

  if (widgets.length <= 0)
    return (
      <Stack textAlign="center">
        <Typography variant="h5" sx={{ textTransform: "initial" }}>
          Sorry, no matching records found
        </Typography>
        <Typography variant="caption" sx={{ textTransform: "initial" }}>
          Add widgets for custom dashboard layout.
        </Typography>
      </Stack>
    );

  console.log(widgets);

  // const layout = [
  //   { widgetID: 1, x: 0, y: 0, w: 1, h: 2 },
  //   { widgetID: 2, x: 1, y: 0, w: 3, h: 2, minW: 2, maxW: 4 },
  //   { widgetID: 3, x: 4, y: 0, w: 1, h: 2 },
  // ];

  const baseLayout = widgets.map((widget, index) => ({
    i: widget.widgetID,
    x: (index % 4) * 12, // 4 per row, spread horizontally
    y: Math.floor(index / 4) * 8, // move vertically after 4 items
    w: 12,
    h: 8,
  }));

  const layouts = {
    lg: baseLayout,
    md: baseLayout,
    sm: baseLayout,
    xs: baseLayout,
    xxs: baseLayout,
  };

  return (
    <>
      <ResponsiveGridLayout
        margin={[20, 20]}
        containerPadding={[10, 10]}
        rowHeight={20}
        layouts={layouts}
        breakpoints={{ lg: 1200, md: 996, sm: 768, xs: 480, xxs: 0 }}
        cols={{ lg: 48, md: 48, sm: 48, xs: 48, xxs: 48 }}
        onLayoutChange={() => {
          console.log("layout has been changed. save it");
        }}
      >
        {widgets.map((widget) => (
          <div key={widget.widgetID}>
            <Widget widget={widget} handleRemoveWidget={handleRemoveWidget} />
          </div>
        ))}
      </ResponsiveGridLayout>
    </>
  );
}
