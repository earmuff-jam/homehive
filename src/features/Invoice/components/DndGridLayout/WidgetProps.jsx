import React from "react";

import { Box } from "@mui/material";
import DetailsTableView from "features/Invoice/components/Widgets/DetailsTableView";
import InvoiceTimelineChart from "features/Invoice/components/Widgets/InvoiceTimelineChart";
import InvoiceTrendsChart from "features/Invoice/components/Widgets/InvoiceTrends";
import ItemTypeFreqChart from "features/Invoice/components/Widgets/ItemTypeFreqChart";

export default function WidgetProps(widget) {
  switch (widget.id) {
    case 1:
      return (
        <Box sx={{ height: "20rem", width: "40rem" }}>
          <InvoiceTimelineChart label={widget.label} caption={widget.caption} />
        </Box>
      );
    case 2:
      return (
        <Box sx={{ height: "20rem", width: "40rem" }}>
          <InvoiceTrendsChart label={widget.label} caption={widget.caption} />
        </Box>
      );
    case 3:
      return (
        <Box sx={{ height: "20rem", width: "40rem" }}>
          <ItemTypeFreqChart label={widget.label} caption={widget.caption} />
        </Box>
      );
    case 4:
      return <DetailsTableView label={widget.label} caption={widget.caption} />;
    default:
      return null;
  }
}
