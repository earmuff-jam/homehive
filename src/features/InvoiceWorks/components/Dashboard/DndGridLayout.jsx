import React, { useState } from "react";

import { DndContext, DragOverlay, pointerWithin } from "@dnd-kit/core";
import {
  SortableContext,
  arrayMove,
  rectSortingStrategy,
} from "@dnd-kit/sortable";
import { Box, Stack, Typography } from "@mui/material";
import Widget from "features/InvoiceWorks/components/Dashboard/Widget";
import WidgetProps from "features/InvoiceWorks/components/Dashboard/WidgetProps";

export default function DndGridLayout({
  editMode,
  widgets,
  setWidgets,
  handleRemoveWidget,
}) {
  const [activeWidget, setActiveWidget] = useState(null); // active widget for drag overlay

  const handleDragStart = (ev) => {
    const { active } = ev;
    const activeId = active.id.toString();
    const widget = widgets.find((w) => w.widgetID === activeId);
    setActiveWidget(widget);
  };

  const handleDragEnd = (ev) => {
    const { active, over } = ev; // active is current, over is replacing

    setActiveWidget(null);

    if (!over) return;
    if (active.id === over.id) return;

    const originalIdx = widgets.findIndex(
      (widget) => widget.widgetID === active.id.toString(),
    );
    const newIdx = widgets.findIndex(
      (widget) => widget.widgetID === over.id.toString(),
    );

    const updatedWidgets = arrayMove(widgets, originalIdx, newIdx);

    setWidgets(updatedWidgets);
    localStorage.setItem("widgets", JSON.stringify(updatedWidgets));
  };

  if (widgets.length <= 0)
    return (
      <Stack textAlign="center">
        <Typography variant="h5">Sorry, no matching records found</Typography>
        <Typography variant="caption">
          Add widgets for custom dashboard layout.
        </Typography>
      </Stack>
    );

  return (
    <DndContext
      collisionDetection={pointerWithin}
      onDragStart={handleDragStart}
      onDragEnd={handleDragEnd}
    >
      <SortableContext
        items={widgets.map((w) => w.widgetID)}
        strategy={rectSortingStrategy}
      >
        <Stack
          sx={{
            display: "flex",
            flexDirection: "row",
            flexWrap: "wrap",
            gap: 2,
            margin: "1rem 0rem",
          }}
        >
          {widgets.map((widget) => {
            const isDragging = activeWidget?.widgetID === widget.widgetID;
            return (
              <Box key={widget.widgetID}>
                {isDragging ? (
                  <Box
                    sx={{
                      width: activeWidget?.config?.width,
                      height: activeWidget?.config?.height,
                      backgroundColor: "slategrey",
                    }}
                  />
                ) : (
                  <Widget
                    widget={widget}
                    editMode={editMode}
                    handleRemoveWidget={handleRemoveWidget}
                  >
                    {WidgetProps(widget)}
                  </Widget>
                )}
              </Box>
            );
          })}
        </Stack>
      </SortableContext>
      <DragOverlay>
        {activeWidget ? (
          <Widget widget={activeWidget} handleRemoveWidget={() => {}} />
        ) : null}
      </DragOverlay>
    </DndContext>
  );
}
