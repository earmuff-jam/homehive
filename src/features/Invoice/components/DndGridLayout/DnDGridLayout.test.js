import React from "react";
import { render, screen } from "@testing-library/react";
import DndGridLayout from "./DndGridLayout";

describe("DndGridLayout", () => {
  const mockSetWidgets = jest.fn();
  const mockHandleRemoveWidget = jest.fn();

  it("renders empty state when no widgets exist", () => {
    const { asFragment } = render(
      <DndGridLayout
        editMode={false}
        widgets={[]}
        setWidgets={mockSetWidgets}
        handleRemoveWidget={mockHandleRemoveWidget}
      />
    );

    expect(screen.getByText(/Sorry, no matching records found/i)).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders widget layout when widgets exist", () => {
    const widgets = [
      {
        widgetID: "1",
        id: "1",
        label: "Test Widget",
        config: { width: 200, height: 100 },
      },
    ];

    const { asFragment } = render(
      <DndGridLayout
        editMode={false}
        widgets={widgets}
        setWidgets={mockSetWidgets}
        handleRemoveWidget={mockHandleRemoveWidget}
      />
    );

    // snapshot to capture structure
    expect(asFragment()).toMatchSnapshot();
  });
});
