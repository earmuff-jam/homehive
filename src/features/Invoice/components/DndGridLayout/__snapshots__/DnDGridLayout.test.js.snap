// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DndGridLayout renders empty state when no widgets exist 1`] = `
<DocumentFragment>
  <div
    class="MuiStack-root css-5andtu-MuiStack-root"
  >
    <h5
      class="MuiTypography-root MuiTypography-h5 css-16xl4zq-MuiTypography-root"
    >
      Sorry, no matching records found
    </h5>
    <span
      class="MuiTypography-root MuiTypography-caption css-mpaqru-MuiTypography-root"
    >
      Add widgets for custom dashboard layout.
    </span>
  </div>
</DocumentFragment>
`;

exports[`DndGridLayout renders widget layout when widgets exist 1`] = `
<DocumentFragment>
  <div
    class="MuiStack-root css-1s5gmop-MuiStack-root"
  >
    <div
      class="MuiBox-root css-0"
    >
      <div
        class="MuiBox-root css-0"
      >
        <span
          class="MuiBadge-root css-63ezxx-MuiBadge-root"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiBox-root css-1ka4bzq-MuiPaper-root"
            style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
          >
            <div
              class="MuiStack-root css-niqf4j-MuiStack-root"
            />
          </div>
          <span
            class="MuiBadge-badge MuiBadge-standard MuiBadge-anchorOriginTopRight MuiBadge-anchorOriginTopRightRectangular MuiBadge-overlapRectangular css-2ao5rv-MuiBadge-badge"
          />
        </span>
      </div>
    </div>
  </div>
  <div
    id="DndDescribedBy-0"
    style="display: none;"
  >
    
    To pick up a draggable item, press the space bar.
    While dragging, use the arrow keys to move the item.
    Press space again to drop the item in its new position, or press escape to cancel.
  
  </div>
  <div
    aria-atomic="true"
    aria-live="assertive"
    id="DndLiveRegion-0"
    role="status"
    style="position: fixed; top: 0px; left: 0px; width: 1px; height: 1px; margin: -1px; border: 0px; padding: 0px; overflow: hidden; clip-path: inset(100%); white-space: nowrap;"
  />
</DocumentFragment>
`;
